pkg/deploy/internal/document/document.go

```diff
@@ -0,0 +1,174 @@
+		if err == nil {
+			return createResolvedEntity(documentName, updateResponse.ID, c.Coordinate, properties), nil
+		}
+
+		if !isAPIErrorStatusNotFound(err) {
+			return entities.ResolvedEntity{}, deployErrors.NewConfigDeployErr(c, fmt.Sprintf("failed to update document '%s'", c.OriginObjectId)).WithError(err)
+		}
+	}
+
+	// strategy 2: find and update document via external id
+	externalId, err := idutils.GenerateExternalIDForDocument(c.Coordinate)
+	if err != nil {
+		return entities.ResolvedEntity{}, deployErrors.NewConfigDeployErr(c, "error generating external id").WithError(err)
+	}
+
+	id, err := tryGetDocumentIDByExternalID(ctx, client, externalId)
+	if err != nil {
+		return entities.ResolvedEntity{}, deployErrors.NewConfigDeployErr(c, fmt.Sprintf("error finding document with externalId='%s'", externalId)).WithError(err)
+	}
+
+	if id != "" {
+		updateResponse, err := client.Update(ctx, id, documentName, []byte(renderedConfig), documentType)
+		if err != nil {
+			return entities.ResolvedEntity{}, deployErrors.NewConfigDeployErr(c, fmt.Sprintf("failed to update document '%s'", c.OriginObjectId)).WithError(err)
+		}
+
+		return createResolvedEntity(documentName, updateResponse.ID, c.Coordinate, properties), nil
+	}
+
+	// strategy 3: try to create a new document
+	createResponse, err := client.Create(ctx, documentName, externalId, []byte(renderedConfig), documentType)
+	if err != nil {
+		return entities.ResolvedEntity{}, deployErrors.NewConfigDeployErr(c, fmt.Sprintf("failed to create document named '%s'", documentName)).WithError(err)
+	}
+
+	return createResolvedEntity(documentName, createResponse.ID, c.Coordinate, properties), nil
+}
+
+func isAPIErrorStatusNotFound(err error) bool {
+	var apiErr api.APIError
+	if !errors.As(err, &apiErr) {
+		return false
+	}
+
+	return apiErr.StatusCode == http.StatusNotFound
+}
+
+func tryGetDocumentIDByExternalID(ctx context.Context, client Client, externalId string) (string, error) {
+	listResponse, err := client.List(ctx, fmt.Sprintf("externalId=='%s'", externalId))
+	if err != nil {
+		return "", err
+	}
+
+	//  it is an error if more than one document was found with the same external id: it should be unique
+	if len(listResponse.Responses) > 1 {
+		return "", fmt.Errorf("multiple documents found with externalId='%s'", externalId)
+	}
+
+	if len(listResponse.Responses) == 0 {
+		return "", nil
+	}
+
+	return listResponse.Responses[0].ID, nil
+}
+
+func createResolvedEntity(documentName string, id string, coordinate coordinate.Coordinate, properties parameter.Properties) entities.ResolvedEntity {
+	properties[config.IdParameter] = id
+
+	return entities.ResolvedEntity{
+		EntityName: documentName,
+		Coordinate: coordinate,
+		Properties: properties,
+	}
+}
+
+func getDocumentTypeFromConfigType(t config.Type) (documents.DocumentType, error) {
+	documentType, ok := t.(config.DocumentType)
```

Comment:
I'm not a fan (and a minority) of repetitive checks - the same check we had before the call.
But if we have it, I would put it as the very first check in the lifecycle of the `Deploy` function.

Reply:
`Deploy` itself doesn't offer much safety so I'd like to leave the check before using the value, but it does make sense to do it earlier

Reply:
nit: I suggest putting this check in the `Deploy` function as the first. To amplify visibility!

Reply:
Not sure I agree here, maybe we should then rework these types though!